### PR TITLE
Neighbours

### PR DESCRIPTION
**Pull request by olofkallander**

### CODE REVIEW

File: test/integration/IntegrationTest.cpp [hunk 1]
```diff
@@ -1440,3 +1440,73 @@ TEST_F(IntegrationTest, conferencePort)
         }
     });
 }
+
+TEST_F(IntegrationTest, neighbours)
```
Comment (olofkallander) [inline test/integration/IntegrationTest.cpp:1444]: I do think we need a BB test for this also.

File: api/Parser.cpp [hunk 2]
```diff
@@ -381,6 +381,13 @@ EndpointDescription parsePatchEndpoint(const nlohmann::json& data, const std::st
         endpointDescription.data.set(dataChannel);
     }
 
+    if (data.find("neighbours") != data.end())
+    {
+        for (auto& group : data["neighbours"]["groups"])
```
Comment (RicardoMDomingues) [inline api/Parser.cpp:386]: Protect against SEGV on BadRequests

```suggestion
        for (auto& group : requiredJsonArray(data["neighbours"]), "groups"))
```

File: bridge/Mixer.cpp [hunk 3]
```diff
@@ -22,6 +22,7 @@
 #include "transport/ice/IceCandidate.h"
 #include "utils/IdGenerator.h"
 #include "utils/SocketAddress.h"
+#include "utils/Span.h"
```
Comment (RicardoMDomingues) [inline bridge/Mixer.cpp:25]: unused?

File: bridge/engine/EngineAudioStream.h [hunk 4]
```diff
@@ -2,6 +2,10 @@
 #include "bridge/RtpMap.h"
 #include "bridge/engine/SsrcOutboundContext.h"
 #include "concurrency/MpmcHashmap.h"
+#include "memory/Map.h"
+#include "utils/FixString.h"
```
Comment (RicardoMDomingues) [inline bridge/engine/EngineAudioStream.h:6]: unused?

File: bridge/engine/EngineAudioStream.h [hunk 5]
```diff
@@ -2,6 +2,10 @@
...
+#include "utils/Span.h"
```
Comment (RicardoMDomingues) [inline bridge/engine/EngineAudioStream.h:7]: unused?

File: bridge/engine/EngineMixer.cpp [hunk 6]
```diff
@@ -2104,6 +2138,19 @@ void EngineMixer::forwardAudioRtpPacket(IncomingPacketInfo& packetInfo, uint64_t
             continue;
         }
 
+        if (srcUserId.isSet() && audioStream->neighbours.contains(srcUserId.get()))
+        {
+            continue;
+        }
+        else if (!audioStream->neighbours.empty())
```
Comment (RicardoMDomingues) [inline bridge/engine/EngineMixer.cpp:2145]:
```suggestion

        if (!audioStream->neighbours.empty())
```

File: bridge/engine/NeighbourMembership.h [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+#include "concurrency/MpmcHashmap.h"
+#include "memory/Array.h"
+#include <cinttypes>
+#include <cstddef>
+
+namespace bridge
+{
+namespace engine
+{
+using NeighbourMembershipArray = memory::Array<uint32_t, 64>;
```
Comment (RicardoMDomingues) [inline bridge/engine/NeighbourMembership.h:11]: I think you should move this to inside of `NeighbourMembership` scope

File: utils/ContainerAlgorithms.h [hunk 8]
```diff
@@ -9,15 +9,28 @@ bool contains(const T& container, UnaryPredicate predicate)
     return std::find_if(container.cbegin(), container.cend(), predicate) != container.cend();
 }
 
+template <typename T, class I>
+bool itemIn(const I& item, const T& container)
```
Comment (RicardoMDomingues) [inline utils/ContainerAlgorithms.h:13]: What is the diference for`std::find`?

**Pull request merged**